Core 559 immutable data structures

dzajkowski opened the pull request:
## Overview
Provide a brief description of what this PR does, and why it's needed.

### Does this PR relate to an RChain JIRA issue? 
If applicable, add link to corresponding JIRA issue.

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [ ] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [ ] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

AbnerZheng committed:
print version when start up rnode

AbnerZheng committed:
Merge branch 'dev' into dev-core524

KentShikama committed:
Fix typo in "missing justifications"

Commit:
[WIP] introduce trie specification tests

tymm committed:
Add opfork and copy method for ctxt

tymm committed:
Fix typo

Commit:
[WIP] common LMDB tests store

lukasz-golebiewski committed:
Add check whether nested transaction's effect is visible in outer

lukasz-golebiewski committed:
Rename insert effect test

lukasz-golebiewski committed:
Add test checking first and second level of trie structure

kjmcqueen committed:
Refactor to protobuf standard style (mostly).
Narrow the detection of self-referencial objects to TupleExpr.

lukasz-golebiewski committed:
Move TestData to HistoryTestsBase

lukasz-golebiewski committed:
[WIP] checke levels after inserting 2 elements

KellyatPyrofex committed:
Delete rholangtut-0.2.md

This version is deprecated. Preserving it in the repo creates user confusion. Please delete. With the launch of RNode v0.4.1, we will make all updates to the rholangtut.md file.

KellyatPyrofex committed:
Delete rholangtut-0.3.md

henrytill committed:
CORE-170: rspace: implement getCheckpoint

lukasz-golebiewski committed:
Refactor four levels test

lukasz-golebiewski committed:
Make txn and store implicit

lukasz-golebiewski committed:
Refactor two levels test

lukasz-golebiewski committed:
Move private method to the bottom

lukasz-golebiewski committed:
Clear ITrieStore before each test

tymm committed:
Add apply method for tuple

lukasz-golebiewski committed:
Add test for failing second delete

Commit:
CORE-559 Replaced mutable maps with immutable var refs.

Commit:
Unit tests for casper sending blocks

Commit:
Fix bug where request for missing block could be repeated

tymm committed:
Use cloneable trait

henrytill committed:
CORE-170: rspace: persist root in db

henrytill committed:
CORE-170: rspace.IStore: remove H, publicize trieStore

henrytill committed:
CORE-170: rspace: add getCheckpoint tests

henrytill committed:
CORE-170: rspace: add a comment and some scaladoc

lukasz-golebiewski committed:
Return false in history.delete if an entry is missing

Commit:
WIP

lukasz-golebiewski committed:
Test retention after delete

Commit:
Merge branch 'CORE-578_align-data-representation' into CORE-559_immutable-data-structures

Commit:
Merge branch 'CORE-578_align-data-representation' into CORE-559_immutable-data-structures

Commit:
Minor clean up of TransportLayerTestImpl

lukasz-golebiewski committed:
Share common data using fixtures

Commit:
removing cloud p2p update on dev from Travis CI

Commit:
Merge branch 'dev' of github.com:rchain/rchain into OPS-143

lukasz-golebiewski committed:
Move common assertions to one place

Commit:
CORE-559 Replace data structures in in-mem store with immutable ones.

tymm committed:
Remove cloneable trait

lukasz-golebiewski commented on rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala:
```diff
@@ -24,10 +23,15 @@ class InMemoryStore[C, P, A, K] private (
 
   val eventsCounter: StoreEventsCounter = new StoreEventsCounter()
 
+  var _dbGNATs: Map[Blake2b256Hash, GNAT[C, P, A, K]] = dbGNATs
```
Why not just do `Map.empty` here and remove the class parameters list? 
BTW, shouldn't the `var` be `private[this]`?

lukasz-golebiewski commented on rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala:
```diff
@@ -24,10 +23,15 @@ class InMemoryStore[C, P, A, K] private (
   val eventsCounter: StoreEventsCounter = new StoreEventsCounter()
 
+  var _dbGNATs: Map[Blake2b256Hash, GNAT[C, P, A, K]] = dbGNATs
+  var _dbJoins: Map[C, Seq[Seq[C]]]                   = dbJoins
```
Same as above

tymm committed:
Merge pull request #837 from tymm/add-ctxt-copy

Add opfork and copy method for ctxt

Commit:
CORE-559 Simplified InMemoryStore creation.

lukasz-golebiewski reviewed (approved):
LGTM

Commit:
CORE-559 bug: missing parens make all the difference

Commit:
Merge pull request #868 from jeremybusk/OPS-143

OPS 143 - Removing P2P test network update from CI

KellyatPyrofex committed:
Merge pull request #857 from rchain/KellyatPyrofex-patch-1

Delete rholangtut-0.2.md

KellyatPyrofex committed:
Merge pull request #858 from rchain/KellyatPyrofex-patch-3

Delete rholangtut-0.3.md

Commit:
introducing gitlab config to CI

lukasz-golebiewski committed:
Merge pull request #862 from lukasz-golebiewski/double-history-delete-is-terrible

Add test for failing second delete

dzajkowski committed:
Merge pull request #861 from dzajkowski/CORE-578_align-data-representation

Core 578 Refactor InMemoryStore to use Blake2b256Hash-> GNAT

KellyatPyrofex committed:
Removed reference to RNode v0.3

kjmcqueen committed:
Merge pull request #846 from kjmcqueen/CompileExportObjects4

Refactor to protobuf standard style (mostly).

Commit:
Merge pull request #870 from jeremybusk/OPS-144

OPS 144 - Adding Gitlab to the CI Mix

Commit:
update name for ci config job

birchmd committed:
Merge pull request #863 from birchmd/dev-birch-casper-unit-tests

Unit tests for casper sending blocks

KentShikama committed:
Add check for equivocations and expected equivocation fetches

KentShikama committed:
Add research/sharding/ve to .gitignore

KentShikama committed:
Merge commit 'c9aa86bcad37721f53976d230e507b8fd7e38777' into dev-kent-casper-equivocations

# Conflicts:
#	casper/src/main/scala/coop/rchain/casper/Casper.scala

henrytill committed:
Merge branch 'dev' into CORE-170-rspace-getCheckpoint

henrytill committed:
Merge pull request #777 from AbnerZheng/dev-core524

CORE-524 : print version when starting up rnode

henrytill committed:
Merge pull request #860 from henrytill/CORE-170-rspace-getCheckpoint

CORE-170: rspace: implement getCheckpoint

KentShikama committed:
Fix false positive in equivocation flagging

We used to compare justification to a last message hash.
We now appropriately compare the latestBlockHash part of
the justification to the last message hash for the sender.

KellyatPyrofex committed:
Merge pull request #872 from rchain/KellyatPyrofex-patch-4

Removed reference to RNode v0.3

Commit:
Merge branch 'dev' into CORE-559_immutable-data-structures

KentShikama committed:
Change addBlocks to use BlockStatus

Insert equivocation case for adding blocks

KentShikama committed:
Remove TODO that asks about capturing effectful code

henrytill committed:
CORE-575: rspace.history: add getLeaves

KentShikama committed:
Add test to confirm fetching of parents even if equivocation

henrytill committed:
Merge pull request #878 from henrytill/CORE-575-rspace-add-getLeaves

CORE-575: rspace.history: add getLeaves

henrytill committed:
CORE-575: rspace.history: restructure a bit

henrytill committed:
CORE-575: rspace: remove ITestableStore

Now that clear needs to be part of the IStore interface, we can
simplify things by removing ITestableStore.

henrytill committed:
CORE-575: rspace.history: tweak ITrieStore#clear

henrytill committed:
CORE-575: rspace: allow withTestStore to return a value

henrytill committed:
CORE-575: rspace: implement reset

henrytill committed:
CORE-575: rspace: update and expand reset tests

henrytill committed:
CORE-575: rspace: tweak reset tests

KentShikama committed:
Refactor SafetyOracle to use BlockHash/Validator

This is more descriptive than ByteString.

KentShikama committed:
Change safety oracle API to return normalized fault tolerance

KentShikama committed:
Refactor createChain to always be a function in casper/tests

lukasz-golebiewski committed:
Merge branch 'dev' of github.com:rchain/rchain into chore/test-trie-structure

KentShikama committed:
Change comment to suggest -1 FT doesn't always mean orphaned

KentShikama committed:
Update addBlock to use BlockStatus

KentShikama committed:
Integrate Turan Oracle into casper log outputs

KentShikama committed:
Update wording of log message for HashSetCasperTest

KentShikama committed:
Change normalizedFaultTolerance to take a blockDag

lukasz-golebiewski committed:
Rename WithLMDBStore to LMDBTrieStoreFixtures

lukasz-golebiewski committed:
Change fixtures to data objects

Commit:
Merge branch 'dev' into CORE-559_immutable-data-structures

Commit:
#CORE-559 Aligned types with generic version.

KentShikama committed:
Address review changes by @birchmd

lukasz-golebiewski committed:
Merge pull request #839 from dzajkowski/chore/test-trie-structure

Test trie structure

henrytill committed:
CORE-575: rspace: add overloaded lookup

henrytill committed:
CORE-575: rspace.history: allElementsOf -> theSameElementsAs

KentShikama committed:
Merge pull request #879 from KentShikama/dev-kent-casper-equivocations

Add check for equivocations and expected equivocation fetches

Commit:
#CORE-559 Applied PR remarks

KentShikama committed:
Merge branch 'dev' into dev-kent-integrate-oracle

henrytill committed:
Merge branch 'dev' into CORE-575-rspace-add-reset

KentShikama committed:
Address review changes by @birchmd

henrytill committed:
CORE-575: rspace: address review comment from @dzajkowski

KentShikama committed:
Rename totalWeight function to computeTotalWeight

henrytill committed:
CORE-575: rspace: tweak history tests

henrytill committed:
Merge pull request #882 from henrytill/CORE-575-rspace-add-reset

CORE-575: rspace: implement reset

KentShikama committed:
Merge pull request #883 from KentShikama/dev-kent-integrate-oracle

Integrate Turan Oracle into Casper logs

Commit:
Merge remote-tracking branch 'upstream/dev' into CORE-559_immutable-data-structures

Commit:
clear def aligned with upstream changes

rchain-travis committed:
Merge pull request #874 from jeremybusk/update-job-name

update name for ci config job

Commit:
Merge branch 'dev' into CORE-559_immutable-data-structures
